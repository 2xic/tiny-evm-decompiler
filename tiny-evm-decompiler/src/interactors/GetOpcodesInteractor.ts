//import { OpCode } from 'tinyeth/dist/evm/OpCode';
import { Opcodes } from 'tinyeth/dist/';
import { OpCode } from 'tinyeth/dist/evm/OpCode';

export class GetOpcodesInteractor {
  public getOpcodes({
    contract
  }: {
    contract: string
  }): ParsedOpcodes[] {
    let contractBuffer = Buffer.from(contract.replace('0x', ''), 'hex')
    const opcodes: ParsedOpcodes[] = [];
    let address = 0;
    let isReadingCode = true;
    let codeBlockType = CodeBlockType.START;

    while (contractBuffer.length) {
      const currentOpcode = contractBuffer[0];
      const opcode: OpCode | undefined = Opcodes[currentOpcode];

      if (opcode?.mnemonic === 'JUMPDEST'){
        isReadingCode = true;
        codeBlockType = CodeBlockType.START;
      }

      if (!opcode) {
        opcodes.push({
          address,
          opcode: 'INVALID',
          value: currentOpcode,
        });
        contractBuffer = contractBuffer.slice(1);
        address += 1;
      } else if (!isReadingCode){
        opcodes.push({
          address,
          opcode: `DATA 0x${currentOpcode.toString(16)}`,
          value: currentOpcode,
        });
        contractBuffer = contractBuffer.slice(1);
      } else {
        if (opcode.isTerminating){
          isReadingCode = false;
          codeBlockType = CodeBlockType.END
        }
        const opcodeArguments = contractBuffer.slice(1, opcode.length);
        opcodes.push({
          address: address,
          opcode: {
            mnemonic: opcode.mnemonic,
            arguments: Array.from(opcodeArguments).map((item) => `0x${item.toString(16)}`),
          },
          value: currentOpcode,
          codeBlockType,
        });
        contractBuffer = contractBuffer.slice(!opcode.length ? 1 : opcode.length);
        address += opcode.length;
      }

      codeBlockType = undefined;
    }

    return opcodes;
  }
}

export type ParsedOpcodes = {
  address: number,
  opcode: string | {
    mnemonic: string;
    arguments: unknown[];
  };
  value: number;
  codeBlockType?: CodeBlockType,
}

export enum CodeBlockType {
  'START',
  'END'
};
