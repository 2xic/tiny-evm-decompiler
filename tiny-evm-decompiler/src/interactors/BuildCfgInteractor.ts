import { GetCodeBlocksInteractor } from "./GetCodeBlocksInteractor";
import { GetControlsFlowInteractor, GraphCodeBlocks } from "./GetControlsFlowInteractor";
import { ParsedOpcodes } from "./GetOpcodesInteractor";
import { ResolveOrphansInteractor } from "./ResolveOrphansInteractor";
import { GetGraphCodeBlockDispatcher } from './dispatcher/GetGraphCodeBlockDispatcher';
import { GetContractFunctionsInteractor } from './dispatcher/GetContractFunctionsInteractor';

export class BuildCfgInteractor {
    public async build({opcodes}: {opcodes: ParsedOpcodes[]}) {
        const contractCodeBlocks = new GetCodeBlocksInteractor().getCodeBlocks(opcodes);
        const codeBlocks = new GetControlsFlowInteractor().getControlFlow({
            codeBlocks: contractCodeBlocks
        });
        const graphBlocks: GraphCodeBlocks[] = new ResolveOrphansInteractor().resolve({
            codeBlocks
        }).map((blocks) => {
            return {
                ...blocks,
            }
        });
        const { graph } = new GetGraphCodeBlockDispatcher().getDispatcherFlow({
            graph: graphBlocks
        })
        
        return {
            graph
        }
    }
}