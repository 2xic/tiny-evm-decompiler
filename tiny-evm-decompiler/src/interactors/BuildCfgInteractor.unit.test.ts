import { BuildCfgInteractor } from "./BuildCfgInteractor";
import { GetOpcodesInteractor } from "./GetOpcodesInteractor";


describe('BuildCfgInteractor', () => {
    it('should create a cfg', async () => {
        const interactor = new BuildCfgInteractor();
        const opcodes = new GetOpcodesInteractor().getOpcodes({
            contract: Buffer.from('6080604052348015600f57600080fd5b506004361060285760003560e01c806312514bba14602d575b600080fd5b603c60383660046076565b603e565b005b60405181815233907f69ca02dd4edd7bf0a4abb9ed3b7af3f14778db5d61921c7dc7cd545266326de29060200160405180910390a250565b600060208284031215608757600080fd5b503591905056fea264697066735822122069fde629581323247d8a4004cb8110511db8008eedae76ad0efb27af7fd3c55464736f6c634300080d0033', 'hex')
        })
        const cfg = await interactor.build({opcodes})
        expect(cfg).toMatchSnapshot();
    })
})