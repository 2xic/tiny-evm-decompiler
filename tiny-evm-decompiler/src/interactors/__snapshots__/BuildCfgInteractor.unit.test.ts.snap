// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BuildCfgInteractor should create a cfg 1`] = `
{
  "graph": [
    {
      "block": [
        {
          "codeBlockType": "START",
          "offset": 0,
          "opcode": {
            "arguments": "0x80",
            "isReal": true,
            "mnemonic": "PUSH1",
            "opcode": 96,
          },
          "value": 96,
        },
        {
          "codeBlockType": undefined,
          "offset": 2,
          "opcode": {
            "arguments": "0x40",
            "isReal": true,
            "mnemonic": "PUSH1",
            "opcode": 96,
          },
          "value": 96,
        },
        {
          "codeBlockType": undefined,
          "offset": 4,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "MSTORE",
            "opcode": 82,
          },
          "value": 82,
        },
        {
          "codeBlockType": undefined,
          "offset": 5,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "CALLVALUE",
            "opcode": 52,
          },
          "value": 52,
        },
        {
          "codeBlockType": undefined,
          "offset": 6,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "DUP1",
            "opcode": 128,
          },
          "value": 128,
        },
        {
          "codeBlockType": undefined,
          "offset": 7,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "ISZERO",
            "opcode": 21,
          },
          "value": 21,
        },
        {
          "codeBlockType": undefined,
          "offset": 8,
          "opcode": {
            "arguments": "0x0f",
            "isReal": true,
            "mnemonic": "PUSH1",
            "opcode": 96,
          },
          "value": 96,
        },
        {
          "codeBlockType": "END",
          "offset": 10,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "JUMPI",
            "opcode": 87,
          },
          "value": 87,
        },
      ],
      "calls": [
        "f",
        "b",
      ],
      "endAddress": 11,
      "isPartOfDispatcher": true,
      "name": "0",
      "properties": [],
      "startAddress": 0,
      "successors": undefined,
    },
    {
      "block": [
        {
          "codeBlockType": undefined,
          "offset": 11,
          "opcode": {
            "arguments": "0x00",
            "isReal": true,
            "mnemonic": "PUSH1",
            "opcode": 96,
          },
          "value": 96,
        },
        {
          "codeBlockType": undefined,
          "offset": 13,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "DUP1",
            "opcode": 128,
          },
          "value": 128,
        },
        {
          "codeBlockType": "END",
          "offset": 14,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "REVERT",
            "opcode": 253,
          },
          "value": 253,
        },
      ],
      "calls": [],
      "endAddress": 15,
      "isPartOfDispatcher": true,
      "name": "b",
      "properties": [],
      "startAddress": 11,
      "successors": undefined,
    },
    {
      "block": [
        {
          "codeBlockType": "START",
          "offset": 15,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "JUMPDEST",
            "opcode": 91,
          },
          "value": 91,
        },
        {
          "codeBlockType": undefined,
          "offset": 16,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "POP",
            "opcode": 80,
          },
          "value": 80,
        },
        {
          "codeBlockType": undefined,
          "offset": 17,
          "opcode": {
            "arguments": "0x04",
            "isReal": true,
            "mnemonic": "PUSH1",
            "opcode": 96,
          },
          "value": 96,
        },
        {
          "codeBlockType": undefined,
          "offset": 19,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "CALLDATASIZE",
            "opcode": 54,
          },
          "value": 54,
        },
        {
          "codeBlockType": undefined,
          "offset": 20,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "LT",
            "opcode": 16,
          },
          "value": 16,
        },
        {
          "codeBlockType": undefined,
          "offset": 21,
          "opcode": {
            "arguments": "0x28",
            "isReal": true,
            "mnemonic": "PUSH1",
            "opcode": 96,
          },
          "value": 96,
        },
        {
          "codeBlockType": "END",
          "offset": 23,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "JUMPI",
            "opcode": 87,
          },
          "value": 87,
        },
      ],
      "calls": [
        "28",
        "18",
      ],
      "endAddress": 24,
      "isPartOfDispatcher": true,
      "name": "f",
      "properties": [
        "JUMPDEST",
      ],
      "startAddress": 15,
      "successors": undefined,
    },
    {
      "block": [
        {
          "codeBlockType": undefined,
          "offset": 24,
          "opcode": {
            "arguments": "0x00",
            "isReal": true,
            "mnemonic": "PUSH1",
            "opcode": 96,
          },
          "value": 96,
        },
        {
          "codeBlockType": undefined,
          "offset": 26,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "CALLDATALOAD",
            "opcode": 53,
          },
          "value": 53,
        },
        {
          "codeBlockType": undefined,
          "offset": 27,
          "opcode": {
            "arguments": "0xe0",
            "isReal": true,
            "mnemonic": "PUSH1",
            "opcode": 96,
          },
          "value": 96,
        },
        {
          "codeBlockType": undefined,
          "offset": 29,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "SHR",
            "opcode": 28,
          },
          "value": 28,
        },
        {
          "codeBlockType": undefined,
          "offset": 30,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "DUP1",
            "opcode": 128,
          },
          "value": 128,
        },
        {
          "codeBlockType": undefined,
          "offset": 31,
          "opcode": {
            "arguments": "0x12514bba",
            "isReal": true,
            "mnemonic": "PUSH4",
            "opcode": 99,
          },
          "value": 99,
        },
        {
          "codeBlockType": undefined,
          "offset": 36,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "EQ",
            "opcode": 20,
          },
          "value": 20,
        },
        {
          "codeBlockType": undefined,
          "offset": 37,
          "opcode": {
            "arguments": "0x2d",
            "isReal": true,
            "mnemonic": "PUSH1",
            "opcode": 96,
          },
          "value": 96,
        },
        {
          "codeBlockType": "END",
          "offset": 39,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "JUMPI",
            "opcode": 87,
          },
          "value": 87,
        },
      ],
      "calls": [
        "2d",
        "28",
      ],
      "endAddress": 40,
      "isPartOfDispatcher": true,
      "name": "18",
      "properties": [],
      "startAddress": 24,
      "successors": undefined,
    },
    {
      "block": [
        {
          "codeBlockType": "START",
          "offset": 40,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "JUMPDEST",
            "opcode": 91,
          },
          "value": 91,
        },
        {
          "codeBlockType": undefined,
          "offset": 41,
          "opcode": {
            "arguments": "0x00",
            "isReal": true,
            "mnemonic": "PUSH1",
            "opcode": 96,
          },
          "value": 96,
        },
        {
          "codeBlockType": undefined,
          "offset": 43,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "DUP1",
            "opcode": 128,
          },
          "value": 128,
        },
        {
          "codeBlockType": "END",
          "offset": 44,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "REVERT",
            "opcode": 253,
          },
          "value": 253,
        },
      ],
      "calls": [],
      "endAddress": 45,
      "isPartOfDispatcher": true,
      "name": "28",
      "properties": [
        "JUMPDEST",
      ],
      "startAddress": 40,
      "successors": undefined,
    },
    {
      "block": [
        {
          "codeBlockType": "START",
          "offset": 45,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "JUMPDEST",
            "opcode": 91,
          },
          "value": 91,
        },
        {
          "codeBlockType": undefined,
          "offset": 46,
          "opcode": {
            "arguments": "0x3c",
            "isReal": true,
            "mnemonic": "PUSH1",
            "opcode": 96,
          },
          "value": 96,
        },
        {
          "codeBlockType": undefined,
          "offset": 48,
          "opcode": {
            "arguments": "0x38",
            "isReal": true,
            "mnemonic": "PUSH1",
            "opcode": 96,
          },
          "value": 96,
        },
        {
          "codeBlockType": undefined,
          "offset": 50,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "CALLDATASIZE",
            "opcode": 54,
          },
          "value": 54,
        },
        {
          "codeBlockType": undefined,
          "offset": 51,
          "opcode": {
            "arguments": "0x04",
            "isReal": true,
            "mnemonic": "PUSH1",
            "opcode": 96,
          },
          "value": 96,
        },
        {
          "codeBlockType": undefined,
          "offset": 53,
          "opcode": {
            "arguments": "0x76",
            "isReal": true,
            "mnemonic": "PUSH1",
            "opcode": 96,
          },
          "value": 96,
        },
        {
          "codeBlockType": "END",
          "offset": 55,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "JUMP",
            "opcode": 86,
          },
          "value": 86,
        },
      ],
      "calls": [
        "76",
      ],
      "endAddress": 56,
      "isPartOfDispatcher": true,
      "name": "2d",
      "properties": [
        "JUMPDEST",
      ],
      "startAddress": 45,
      "successors": undefined,
    },
    {
      "block": [
        {
          "codeBlockType": "START",
          "offset": 56,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "JUMPDEST",
            "opcode": 91,
          },
          "value": 91,
        },
        {
          "codeBlockType": undefined,
          "offset": 57,
          "opcode": {
            "arguments": "0x3e",
            "isReal": true,
            "mnemonic": "PUSH1",
            "opcode": 96,
          },
          "value": 96,
        },
        {
          "codeBlockType": "END",
          "offset": 59,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "JUMP",
            "opcode": 86,
          },
          "value": 86,
        },
      ],
      "calls": [
        "3e",
      ],
      "endAddress": 60,
      "isPartOfDispatcher": true,
      "name": "38",
      "properties": [
        "JUMPDEST",
      ],
      "startAddress": 56,
      "successors": undefined,
    },
    {
      "block": [
        {
          "codeBlockType": "START",
          "offset": 60,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "JUMPDEST",
            "opcode": 91,
          },
          "value": 91,
        },
        {
          "codeBlockType": "END",
          "offset": 61,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "STOP",
            "opcode": 0,
          },
          "value": 0,
        },
      ],
      "calls": [],
      "endAddress": 62,
      "isPartOfDispatcher": false,
      "name": "3c",
      "properties": [
        "JUMPDEST",
        "STOP",
      ],
      "startAddress": 60,
      "successors": undefined,
    },
    {
      "block": [
        {
          "codeBlockType": "START",
          "offset": 62,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "JUMPDEST",
            "opcode": 91,
          },
          "value": 91,
        },
        {
          "codeBlockType": undefined,
          "offset": 63,
          "opcode": {
            "arguments": "0x40",
            "isReal": true,
            "mnemonic": "PUSH1",
            "opcode": 96,
          },
          "value": 96,
        },
        {
          "codeBlockType": undefined,
          "offset": 65,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "MLOAD",
            "opcode": 81,
          },
          "value": 81,
        },
        {
          "codeBlockType": undefined,
          "offset": 66,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "DUP2",
            "opcode": 129,
          },
          "value": 129,
        },
        {
          "codeBlockType": undefined,
          "offset": 67,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "DUP2",
            "opcode": 129,
          },
          "value": 129,
        },
        {
          "codeBlockType": undefined,
          "offset": 68,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "MSTORE",
            "opcode": 82,
          },
          "value": 82,
        },
        {
          "codeBlockType": undefined,
          "offset": 69,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "CALLER",
            "opcode": 51,
          },
          "value": 51,
        },
        {
          "codeBlockType": undefined,
          "offset": 70,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "SWAP1",
            "opcode": 144,
          },
          "value": 144,
        },
        {
          "codeBlockType": undefined,
          "offset": 71,
          "opcode": {
            "arguments": "0x69ca02dd4edd7bf0a4abb9ed3b7af3f14778db5d61921c7dc7cd545266326de2",
            "isReal": true,
            "mnemonic": "PUSH32",
            "opcode": 127,
          },
          "value": 127,
        },
        {
          "codeBlockType": undefined,
          "offset": 104,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "SWAP1",
            "opcode": 144,
          },
          "value": 144,
        },
        {
          "codeBlockType": undefined,
          "offset": 105,
          "opcode": {
            "arguments": "0x20",
            "isReal": true,
            "mnemonic": "PUSH1",
            "opcode": 96,
          },
          "value": 96,
        },
        {
          "codeBlockType": undefined,
          "offset": 107,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "ADD",
            "opcode": 1,
          },
          "value": 1,
        },
        {
          "codeBlockType": undefined,
          "offset": 108,
          "opcode": {
            "arguments": "0x40",
            "isReal": true,
            "mnemonic": "PUSH1",
            "opcode": 96,
          },
          "value": 96,
        },
        {
          "codeBlockType": undefined,
          "offset": 110,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "MLOAD",
            "opcode": 81,
          },
          "value": 81,
        },
        {
          "codeBlockType": undefined,
          "offset": 111,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "DUP1",
            "opcode": 128,
          },
          "value": 128,
        },
        {
          "codeBlockType": undefined,
          "offset": 112,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "SWAP2",
            "opcode": 145,
          },
          "value": 145,
        },
        {
          "codeBlockType": undefined,
          "offset": 113,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "SUB",
            "opcode": 3,
          },
          "value": 3,
        },
        {
          "codeBlockType": undefined,
          "offset": 114,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "SWAP1",
            "opcode": 144,
          },
          "value": 144,
        },
        {
          "codeBlockType": undefined,
          "offset": 115,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "LOG2",
            "opcode": 162,
          },
          "value": 162,
        },
        {
          "codeBlockType": undefined,
          "offset": 116,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "POP",
            "opcode": 80,
          },
          "value": 80,
        },
        {
          "codeBlockType": "END",
          "offset": 117,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "JUMP",
            "opcode": 86,
          },
          "value": 86,
        },
      ],
      "calls": [
        "3c",
      ],
      "endAddress": 118,
      "isPartOfDispatcher": false,
      "name": "3e",
      "properties": [
        "JUMPDEST",
        "LOG",
      ],
      "startAddress": 62,
      "successors": undefined,
    },
    {
      "block": [
        {
          "codeBlockType": "START",
          "offset": 118,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "JUMPDEST",
            "opcode": 91,
          },
          "value": 91,
        },
        {
          "codeBlockType": undefined,
          "offset": 119,
          "opcode": {
            "arguments": "0x00",
            "isReal": true,
            "mnemonic": "PUSH1",
            "opcode": 96,
          },
          "value": 96,
        },
        {
          "codeBlockType": undefined,
          "offset": 121,
          "opcode": {
            "arguments": "0x20",
            "isReal": true,
            "mnemonic": "PUSH1",
            "opcode": 96,
          },
          "value": 96,
        },
        {
          "codeBlockType": undefined,
          "offset": 123,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "DUP3",
            "opcode": 130,
          },
          "value": 130,
        },
        {
          "codeBlockType": undefined,
          "offset": 124,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "DUP5",
            "opcode": 132,
          },
          "value": 132,
        },
        {
          "codeBlockType": undefined,
          "offset": 125,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "SUB",
            "opcode": 3,
          },
          "value": 3,
        },
        {
          "codeBlockType": undefined,
          "offset": 126,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "SLT",
            "opcode": 18,
          },
          "value": 18,
        },
        {
          "codeBlockType": undefined,
          "offset": 127,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "ISZERO",
            "opcode": 21,
          },
          "value": 21,
        },
        {
          "codeBlockType": undefined,
          "offset": 128,
          "opcode": {
            "arguments": "0x87",
            "isReal": true,
            "mnemonic": "PUSH1",
            "opcode": 96,
          },
          "value": 96,
        },
        {
          "codeBlockType": "END",
          "offset": 130,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "JUMPI",
            "opcode": 87,
          },
          "value": 87,
        },
      ],
      "calls": [
        "87",
        "83",
      ],
      "endAddress": 131,
      "isPartOfDispatcher": false,
      "name": "76",
      "properties": [
        "JUMPDEST",
      ],
      "startAddress": 118,
      "successors": undefined,
    },
    {
      "block": [
        {
          "codeBlockType": undefined,
          "offset": 131,
          "opcode": {
            "arguments": "0x00",
            "isReal": true,
            "mnemonic": "PUSH1",
            "opcode": 96,
          },
          "value": 96,
        },
        {
          "codeBlockType": undefined,
          "offset": 133,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "DUP1",
            "opcode": 128,
          },
          "value": 128,
        },
        {
          "codeBlockType": "END",
          "offset": 134,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "REVERT",
            "opcode": 253,
          },
          "value": 253,
        },
      ],
      "calls": [],
      "endAddress": 135,
      "isPartOfDispatcher": false,
      "name": "83",
      "properties": [],
      "startAddress": 131,
      "successors": undefined,
    },
    {
      "block": [
        {
          "codeBlockType": "START",
          "offset": 135,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "JUMPDEST",
            "opcode": 91,
          },
          "value": 91,
        },
        {
          "codeBlockType": undefined,
          "offset": 136,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "POP",
            "opcode": 80,
          },
          "value": 80,
        },
        {
          "codeBlockType": undefined,
          "offset": 137,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "CALLDATALOAD",
            "opcode": 53,
          },
          "value": 53,
        },
        {
          "codeBlockType": undefined,
          "offset": 138,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "SWAP2",
            "opcode": 145,
          },
          "value": 145,
        },
        {
          "codeBlockType": undefined,
          "offset": 139,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "SWAP1",
            "opcode": 144,
          },
          "value": 144,
        },
        {
          "codeBlockType": undefined,
          "offset": 140,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "POP",
            "opcode": 80,
          },
          "value": 80,
        },
        {
          "codeBlockType": "END",
          "offset": 141,
          "opcode": {
            "arguments": "",
            "isReal": true,
            "mnemonic": "JUMP",
            "opcode": 86,
          },
          "value": 86,
        },
      ],
      "calls": [
        "38",
      ],
      "endAddress": 142,
      "isPartOfDispatcher": false,
      "name": "87",
      "properties": [
        "JUMPDEST",
      ],
      "startAddress": 135,
      "successors": undefined,
    },
  ],
}
`;
